frontend/pages/admin/IntegrationsPage/cards/MdmSettings/components/EndUserMigrationSection/EndUserMigrationSection.tsx

```diff
@@ -0,0 +1,217 @@
+import React, { useContext, useState } from "react";
+import { InjectedRouter } from "react-router";
+import classnames from "classnames";
+
+import PATHS from "router/paths";
+import configAPI from "services/entities/config";
+import { AppContext } from "context/app";
+import { NotificationContext } from "context/notification";
+
+// @ts-ignore
+import InputField from "components/forms/fields/InputField";
+import Radio from "components/forms/fields/Radio/Radio";
+import Slider from "components/forms/fields/Slider/Slider";
+import Button from "components/buttons/Button/Button";
+import validateUrl from "components/forms/validators/valid_url";
+import PremiumFeatureMessage from "components/PremiumFeatureMessage/PremiumFeatureMessage";
+import EmptyTable from "components/EmptyTable/EmptyTable";
+
+import ExampleWebhookUrlPayloadModal from "../ExampleWebhookUrlPayloadModal/ExampleWebhookUrlPayloadModal";
+
+const baseClass = "end-user-migration-section";
+
+const VOLUNTARY_MODE_DESCRIPTION =
+  "The end user sees the above window when they select Migrate to Fleet in the Fleet Desktop menu. If they’re unenrolled from your old MDM, the window appears every 15 minutes.";
+const FORCED_MODE_DESCRIPTION =
+  "The end user sees the above window every 15 minutes.";
+
+interface IEndUserMigrationFormData {
+  isEnabled: boolean;
+  mode: "voluntary" | "forced";
+  webhookUrl: string;
+}
+
+interface IEndUserMigrationSectionProps {
+  router: InjectedRouter;
+}
+
+const validateWebhookUrl = (val: string) => {
+  return validateUrl({ url: val });
+};
+
+const EndUserMigrationSection = ({ router }: IEndUserMigrationSectionProps) => {
+  const { config, isPremiumTier, setConfig } = useContext(AppContext);
+  const { renderFlash } = useContext(NotificationContext);
+  const [formData, setFormData] = useState<IEndUserMigrationFormData>({
+    isEnabled: config?.mdm.macos_migration.enable || false,
+    mode: config?.mdm.macos_migration.mode || "voluntary",
+    webhookUrl: config?.mdm.macos_migration.webhook_url || "",
+  });
+  const [showExamplePayload, setShowExamplePayload] = useState(false);
+
+  // we only validate this one input so just going to use simple boolean to
+  // track validation. If we need to validate more inputs in the future we can
+  // use a formErrors object.
+  const [isValidWebhookUrl, setIsValidWebhookUrl] = useState(true);
+
+  const toggleExamplePayloadModal = () => {
+    setShowExamplePayload(!showExamplePayload);
+  };
+
+  const toggleMigrationEnabled = () => {
+    setFormData({ ...formData, isEnabled: !formData.isEnabled });
```

Comment:
I'm not sure that it makes a difference in this particular case, but maybe we want to use the updater function pattern with form data objects?

```
setFormData((prevFormData) => ({ ...prevFormData, isEnabled: !prevFormData.isEnabled });
```

Reply:
right. im not sure this is needed here but will add to be safe.